pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +161,145 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
+	ProxyCommand {{.GHExec}} cs ssh -c {{.Name}} --stdio
+	UserKnownHostsFile=/dev/null
+	StrictHostKeyChecking no
+	LogLevel quiet
+	ControlMaster auto
+
+`)
+	if err != nil {
+		return fmt.Errorf("error formatting template: %w", err)
+	}
+
+	ghexec, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	// store a mapping of repository -> remote ssh username. This is
+	// necessary because the username can vary between codespaces, but
+	// since fetching it is slow, we store it here so we at least only do
+	// it once per repository.
+	sshUsers := map[string]string{}
+
+	var status error
+	for _, cs := range codespaces {
+
+		if cs.State != "Available" {
+			fmt.Fprintf(os.Stderr, "skipping unavailable codespace %s: %s\n", cs.Name, cs.State)
+			status = cmdutil.SilentError
+			continue
+		}
+
+		sshUser, ok := sshUsers[cs.Repository.FullName]
+		if !ok {
+			session, err := openSSHSession(ctx, a, cs, nil)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error connecting to codespace: %v\n", err)
+
+				// Move on to the next codespace. We don't want to bail here - just because we're not
+				// able to set up connectivity to one doesn't mean we shouldn't make a best effort to
+				// generate configs for the rest of them.
+				status = cmdutil.SilentError
+				continue
+			}
+			defer session.Close()
+
+			a.StartProgressIndicatorWithLabel(fmt.Sprintf("Fetching SSH Details for %s", cs.Name))
+			_, sshUser, err = session.StartSSHServer(ctx)
+			a.StopProgressIndicator()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error getting ssh server details: %v", err)
+				status = cmdutil.SilentError
+				continue // see above
+			}
+			sshUsers[cs.Repository.FullName] = sshUser
+		}
+
+		conf := codespaceSSHConfig{
+			Name:       cs.Name,
+			EscapedRef: strings.ReplaceAll(cs.GitStatus.Ref, "/", "-"),
+			SSHUser:    sshUser,
+			GHExec:     ghexec,
+		}
+		if err := t.Execute(a.io.Out, conf); err != nil {
+			return err
+		}
+	}
+
+	return status
+}
+
+// codespaceSSHConfig contains values needed to write an OpenSSH host
+// configuration for a single codespace. For example:
+//
+// Host {{Name}}.{{EscapedRef}
+// User {{SSHUser}
+// ProxyCommand {{GHExec}} cs ssh -c {{Name}} --stdio
+//
+// EscapedRef is included in the name to help distinguish between codespaces
+// when tab-completing ssh hostnames. '/' characters in EscapedRef are
+// flattened to '-' to prevent problems with tab completion or when the
+// hostname appears in ControlMaster socket paths.
+type codespaceSSHConfig struct {
+	Name       string // the codespace name, passed to `ssh -c`
+	EscapedRef string // the currently checked-out branch
+	SSHUser    string // the remote ssh username
+	GHExec     string // path used for invoking the current `gh` binary
+}
+
+func openSSHSession(ctx context.Context, a *App, codespace *api.Codespace, liveshareLogger *log.Logger) (*liveshare.Session, error) {
+	// TODO(josebalius): We can fetch the user in parallel to everything else
```

Comment:
I think this TODO can be done by simply moving the GetUser call into the goroutine below, which should improve latency by ~250ms in both the new and existing commands.

Reply:
One second thoughts, the (expensive) GetUser call is loop-invariant. Let's hoist it out of the loop and pass in the user.Login string as a parameter here.

(More efficient still, though more complex, would be to accept a login string parameter, precomputed in the 'config' case, but to call GetUser from within the checkAuthorizedKeys goroutine if login="", for the 'ssh' case, so that it is computed exactly once in both cases, and off the critical path in the ssh case.

Reply:
I didn't approach it exactly like that, but https://github.com/cli/cli/pull/4915/commits/61823997c228ecf8b52e2c1c9fcfdff8f1589549 makes in so that in all cases we `GetUser` exactly once, and in parallel with other i/o.